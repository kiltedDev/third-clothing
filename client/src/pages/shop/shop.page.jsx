import React, { useEffect, lazy, Suspense } from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

import Spinner from '../../components/spinner/spinner.component';
import ErrorBoundary from '../../components/error-boundary/error-boundary.component';

import { fetchSectionsStart } from '../../redux/shop/shop.actions';
import { hideCart } from '../../redux/cart/cart.actions';

import { ShopPageContainer } from './shop.styles';

const SectionsOverview = lazy( () => import( '../../components/sections-overview/sections-overview.component') );
const SectionPageContainer = lazy( () => import( '../section/section.container') );

export const ShopPage = ({ match, fetchSectionsStart, hideCart }) => {
  useEffect( () => {
    hideCart();
    fetchSectionsStart();
  }, [ fetchSectionsStart, hideCart ]);

  return (
    <main className='shop-page'>
      <ErrorBoundary>
        <Suspense fallback={ <Spinner /> }>
          <Route
            exact
            path={ `${match.path}` }
            component={ SectionsOverview }
            />
          <Route
            exact
            path={ `${match.path}/:sectionId` }
            component={ SectionPageContainer }
            />
        </Suspense>
      </ErrorBoundary>
    </main>
  );
}

const mapDispatchToProps = dispatch => ({
  fetchSectionsStart: () => dispatch( fetchSectionsStart() ),
  hideCart: () => dispatch( hideCart() )

})

export default connect( null, mapDispatchToProps )( ShopPage );
