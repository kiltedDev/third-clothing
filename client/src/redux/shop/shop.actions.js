import ShopActionTypes from './shop.types';

import {
  firestore,
  convertCollectionSnapshotToMap
} from '../../firebase/firebase.utils';

export const fetchSectionsStart = () => ({
  type: ShopActionTypes.FETCH_SECTIONS_START
});

export const fetchSectionsSuccess = sectionsMap => ({
  type: ShopActionTypes.FETCH_SECTIONS_SUCCESS,
  payload: sectionsMap
});

export const fetchSectionsFailure = errorMessage => ({
  type: ShopActionTypes.FETCH_SECTIONS_FAILURE,
  payload: errorMessage
});

export const fetchSectionsStartAsync = () => {
  return dispatch => {
    const sectionRef = firestore.collection('sections');
    dispatch(fetchSectionsStart());

    sectionRef
      .get()
      .then(snapshot => {
        const sectionsMap = convertCollectionSnapshotToMap( snapshot );
        dispatch( fetchSectionsSuccess( sectionsMap ) );
      })
      .catch(error => dispatch(fetchSectionsFailure(error.message)));
  };
};
