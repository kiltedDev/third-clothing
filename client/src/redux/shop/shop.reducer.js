// diverged from course here because in a RW scenario we'd be importing
// from a database and I'd like to have this structured more in that
// direction.  p sure they'll do that later anyhow.


import ShopActionTypes from './shop.types'

const INITIAL_STATE = {
  sections: null,
  isFetching: false,
  errorMessage: undefined
};

const shopReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    // NOTE: replaced with Thunk
    case ShopActionTypes.UPDATE_SECTIONS:
      return {
        ...state,
        sections: action.payload
      }
    case ShopActionTypes.FETCH_SECTIONS_START:
      return {
        ...state,
        isFetching: true
      }
    case ShopActionTypes.FETCH_SECTIONS_SUCCESS:
      return {
        ...state,
        isFetching: false,
        sections: action.payload
      }
    case ShopActionTypes.FETCH_SECTIONS_FAILURE:
      return {
        ...state,
        isFetching: false,
        errorMessage: action.payload
      }
    default:
      return state;
  }
}

export default shopReducer;
