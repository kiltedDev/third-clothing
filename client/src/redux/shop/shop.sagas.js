import { all, call, put, takeLatest } from 'redux-saga/effects';
import {
  firestore,
  convertCollectionSnapshotToMap
} from '../../firebase/firebase.utils';
import {
  fetchSectionsSuccess,
  fetchSectionsFailure
} from './shop.actions';

import ShopActionTypes from './shop.types';

export function* fetchSectionsAsync() {
  try {
    const sectionRef = firestore.collection('sections');
    const snapshot = yield sectionRef.get();
    const sectionsMap = yield call(
      convertCollectionSnapshotToMap,
      snapshot
    );
    yield put( fetchSectionsSuccess( sectionsMap ) );
  } catch ( error ) {
    yield put( fetchSectionsFailure( error.message ) );
  }
}

export function* fetchSectionsStart() {
  yield takeLatest(
    ShopActionTypes.FETCH_SECTIONS_START,
    fetchSectionsAsync
  )
}

export function* shopSagas() {
  yield all([
    call( fetchSectionsStart )
  ])
}
