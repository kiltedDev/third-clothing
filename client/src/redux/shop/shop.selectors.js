import { createSelector } from 'reselect';

const selectShop = state => state.shop;

export const selectShopSections = createSelector(
  [selectShop],
  shop => shop.sections
);

export const selectSectionsForPreview = createSelector(
  [selectShopSections],
  sections => sections
  ? Object.keys( sections )
    .map(key =>
      sections[key]
    )
  : []
);

export const selectSection = sectionUrlParam => createSelector(
  [selectShopSections],
  sections => ( sections
  ? sections[sectionUrlParam]
  : null )
)

export const selectIsSectionFetching = createSelector(
  [selectShop],
  shop => shop.isFetching
)

export const selectIsSectionLoaded = createSelector(
  [selectShop],
  shop => !!shop.sections
)
