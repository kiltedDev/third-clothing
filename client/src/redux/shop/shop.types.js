const ShopActionTypes = {
  FETCH_SECTIONS_START: 'FETCH_SECTIONS_START',
  FETCH_SECTIONS_SUCCESS: 'FETCH_SECTIONS_SUCCESS',
  FETCH_SECTIONS_FAILURE: 'FETCH_SECTIONS_FAILURE'
}

export default ShopActionTypes;
