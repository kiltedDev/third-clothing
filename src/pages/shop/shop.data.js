const SHOP_DATA = [
  {
    id: 1,
    title: 'Hats',
    routeName: 'hats',
    items: [
      {
        id: 'hat-1',
        name: 'Brown Brim',
        imageUrl: 'https://i.ibb.co/ZYW3VTp/brown-brim.png',
        price: 25
      },
      {
        id: 'hat-2',
        name: 'Blue Beanie',
        imageUrl: 'https://i.ibb.co/ypkgK0X/blue-beanie.png',
        price: 18
      },
      {
        id: 'hat-3',
        name: 'Brown Cowboy',
        imageUrl: 'https://i.ibb.co/QdJwgmp/brown-cowboy.png',
        price: 35
      },
      {
        id: 'hat-4',
        name: 'Grey Brim',
        imageUrl: 'https://i.ibb.co/RjBLWxB/grey-brim.png',
        price: 25
      },
      {
        id: 'hat-5',
        name: 'Green Beanie',
        imageUrl: 'https://i.ibb.co/YTjW3vF/green-beanie.png',
        price: 18
      },
      {
        id: 'hat-6',
        name: 'Palm Tree Cap',
        imageUrl: 'https://i.ibb.co/rKBDvJX/palm-tree-cap.png',
        price: 14
      },
      {
        id: 'hat-7',
        name: 'Red Beanie',
        imageUrl: 'https://i.ibb.co/bLB646Z/red-beanie.png',
        price: 18
      },
      {
        id: 'hat-8',
        name: 'Wolf Cap',
        imageUrl: 'https://i.ibb.co/1f2nWMM/wolf-cap.png',
        price: 14
      },
      {
        id: 'hat-9',
        name: 'Blue Snapback',
        imageUrl: 'https://i.ibb.co/X2VJP2W/blue-snapback.png',
        price: 16
      }
    ]
  },
  {
    id: 2,
    title: 'Sneakers',
    routeName: 'sneakers',
    items: [
      {
        id: 'sneaker-1',
        name: 'Adidas NMD',
        imageUrl: 'https://i.ibb.co/0s3pdnc/adidas-nmd.png',
        price: 220
      },
      {
        id: 'sneaker-2',
        name: 'Adidas Yeezy',
        imageUrl: 'https://i.ibb.co/dJbG1cT/yeezy.png',
        price: 280
      },
      {
        id: 'sneaker-3',
        name: 'Black Converse',
        imageUrl: 'https://i.ibb.co/bPmVXyP/black-converse.png',
        price: 110
      },
      {
        id: 'sneaker-4',
        name: 'Nike White AirForce',
        imageUrl: 'https://i.ibb.co/1RcFPk0/white-nike-high-tops.png',
        price: 160
      },
      {
        id: 'sneaker-5',
        name: 'Nike Red High Tops',
        imageUrl: 'https://i.ibb.co/QcvzydB/nikes-red.png',
        price: 160
      },
      {
        id: 'sneaker-6',
        name: 'Nike Brown High Tops',
        imageUrl: 'https://i.ibb.co/fMTV342/nike-brown.png',
        price: 160
      },
      {
        id: 'sneaker-7',
        name: 'Air Jordan Limited',
        imageUrl: 'https://i.ibb.co/w4k6Ws9/nike-funky.png',
        price: 190
      },
      {
        id: 'sneaker-8',
        name: 'Timberlands',
        imageUrl: 'https://i.ibb.co/Mhh6wBg/timberlands.png',
        price: 200
      }
    ]
  },
  {
    id: 3,
    title: 'Jackets',
    routeName: 'jackets',
    items: [
      {
        id: 'jacket-1',
        name: 'Black Jean Shearling',
        imageUrl: 'https://i.ibb.co/XzcwL5s/black-shearling.png',
        price: 125
      },
      {
        id: 'jacket-2',
        name: 'Blue Jean Jacket',
        imageUrl: 'https://i.ibb.co/mJS6vz0/blue-jean-jacket.png',
        price: 90
      },
      {
        id: 'jacket-3',
        name: 'Grey Jean Jacket',
        imageUrl: 'https://i.ibb.co/N71k1ML/grey-jean-jacket.png',
        price: 90
      },
      {
        id: 'jacket-4',
        name: 'Brown Shearling',
        imageUrl: 'https://i.ibb.co/s96FpdP/brown-shearling.png',
        price: 165
      },
      {
        id: 'jacket-5',
        name: 'Tan Trench',
        imageUrl: 'https://i.ibb.co/M6hHc3F/brown-trench.png',
        price: 185
      }
    ]
  },
  {
    id: 4,
    title: 'Womens',
    routeName: 'womens',
    items: [
      {
        id: 'women-1',
        name: 'Blue Tanktop',
        imageUrl: 'https://i.ibb.co/7CQVJNm/blue-tank.png',
        price: 25
      },
      {
        id: 'women-2',
        name: 'Floral Blouse',
        imageUrl: 'https://i.ibb.co/4W2DGKm/floral-blouse.png',
        price: 20
      },
      {
        id: 'women-3',
        name: 'Floral Dress',
        imageUrl: 'https://i.ibb.co/KV18Ysr/floral-skirt.png',
        price: 80
      },
      {
        id: 'women-4',
        name: 'Red Dots Dress',
        imageUrl: 'https://i.ibb.co/N3BN1bh/red-polka-dot-dress.png',
        price: 80
      },
      {
        id: 'women-5',
        name: 'Striped Sweater',
        imageUrl: 'https://i.ibb.co/KmSkMbH/striped-sweater.png',
        price: 45
      },
      {
        id: 'women-6',
        name: 'Yellow Track Suit',
        imageUrl: 'https://i.ibb.co/v1cvwNf/yellow-track-suit.png',
        price: 135
      },
      {
        id: 'women-7',
        name: 'White Blouse',
        imageUrl: 'https://i.ibb.co/qBcrsJg/white-vest.png',
        price: 20
      }
    ]
  },
  {
    id: 5,
    title: 'Mens',
    routeName: 'mens',
    items: [
      {
        id: 'men-1',
        name: 'Camo Down Vest',
        imageUrl: 'https://i.ibb.co/xJS0T3Y/camo-vest.png',
        price: 325
      },
      {
        id: 'men-2',
        name: 'Floral T-shirt',
        imageUrl: 'https://i.ibb.co/qMQ75QZ/floral-shirt.png',
        price: 20
      },
      {
        id: 'men-3',
        name: 'Black & White Longsleeve',
        imageUrl: 'https://i.ibb.co/55z32tw/long-sleeve.png',
        price: 25
      },
      {
        id: 'men-4',
        name: 'Pink T-shirt',
        imageUrl: 'https://i.ibb.co/RvwnBL8/pink-shirt.png',
        price: 25
      },
      {
        id: 'men-5',
        name: 'Jean Long Sleeve',
        imageUrl: 'https://i.ibb.co/VpW4x5t/roll-up-jean-shirt.png',
        price: 40
      },
      {
        id: 'men-6',
        name: 'Burgundy T-shirt',
        imageUrl: 'https://i.ibb.co/mh3VM1f/polka-dot-shirt.png',
        price: 25
      }
    ]
  }
]

export default SHOP_DATA;
