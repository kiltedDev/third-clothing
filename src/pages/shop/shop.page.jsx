import React from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

import CollectionsOverviewContainer from '../../components/collections-overview/collections-overview.container';
import CollectionPageContainer from '../collection/collection.container';
import { fetchCollectionsStartAsync } from '../../redux/shop/shop.actions';
import { hideCart } from '../../redux/cart/cart.actions';

class ShopPage extends React.Component {
  componentDidMount() {
    const { fetchCollectionsStartAsync, hideCart } = this.props;
    hideCart();
    fetchCollectionsStartAsync();
  }


  render() {
    const { match } = this.props;
    return (
      <div className='shop-page'>
        <Route
          exact
          path={ `${match.path}` }
          component={ CollectionsOverviewContainer }
        />
        <Route
          exact
          path={ `${match.path}/:collectionId` }
          component={ CollectionPageContainer }
        />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  fetchCollectionsStartAsync: () => dispatch( fetchCollectionsStartAsync() ),
  hideCart: () => dispatch( hideCart() )

})

export default connect( null, mapDispatchToProps )( ShopPage );
