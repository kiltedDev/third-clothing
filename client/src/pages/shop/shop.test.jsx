import { mount } from 'enzyme';
import React from 'react';
import { combineReducers, createStore } from 'redux';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { ShopPage } from './shop.page';

const createMockStore = ({ state, reducers }) => {
  const store = createStore(combineReducers(reducers), state);
  return {
    ...store,
    persistor: {
      persist: () => null
    }
  };
};

describe('ShopPage', () => {
  let wrapper;
  let snapshot;
  let mockMatch;
  let mockFetchSectionsStart;
  let mockHideCart;
  let store;
  beforeAll(() => {
    const mockReducer = (
      state = {
        isFetching: true
      },
      action
    ) => state;

    const mockState = {
      shop: {
        isFetching: true
      }
    }

    mockMatch = jest.fn();
    mockFetchSectionsStart = jest.fn();
    mockHideCart = jest.fn();

    store = createMockStore({
      state: mockState,
      reducers: { shop: mockReducer }
    });

    const mockProps = {
      match: mockMatch,
      hideCart: mockHideCart,
      fetchSectionsStart: mockFetchSectionsStart
    };
    wrapper = mount(
      <BrowserRouter>
        <Provider store={ store }>
          <ShopPage { ...mockProps } />
        </Provider>
      </BrowserRouter>
    );
    snapshot = wrapper.debug();
  });

  it('should render ShopPage component',() => {
    expect( snapshot ).toMatchSnapshot();
  });

  it('should call fetchSectionsStart upon mounting',() => {
    expect( mockFetchSectionsStart ).toHaveBeenCalled();
  });

  it('should call hideCart upon mounting',() => {
    expect( mockHideCart ).toHaveBeenCalled();
  });
});
