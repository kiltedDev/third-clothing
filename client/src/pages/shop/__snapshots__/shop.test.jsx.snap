// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage should render ShopPage component 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <Provider store={{...}}>
      <ShopPage match={[Function: mockConstructor]} hideCart={[Function: mockConstructor]} fetchSectionsStart={[Function: mockConstructor]}>
        <main className=\\"shop-page\\">
          <ErrorBoundary>
            <Suspense fallback={{...}}>
              <Route exact={true} path=\\"undefined\\" component={{...}} />
              <Route exact={true} path=\\"undefined/:sectionId\\" component={{...}} />
            </Suspense>
          </ErrorBoundary>
        </main>
      </ShopPage>
    </Provider>
  </Router>
</BrowserRouter>"
`;
